topdown/eval_test.go

```diff
@@ -239,3 +243,100 @@ func TestContainsNestedRefOrCall(t *testing.T) {
 		})
 	}
 }
+
+func TestTopdownVirtualCacheFunctions(t *testing.T) {
+	ctx := context.Background()
+	store := inmem.New()
+	compiler := compileModules([]string{
+		`package p
+
+		f(0) = 1
+		f(x) = 12 { x > 0 }
+
+		g(x) = true { x > 0 }
+		g(x) = false { x <= 0 }
+
+		q = y {
+			x := f(1)
+			y := f(1)
+			x == y
+		}
+		`,
+	})
+
+	tests := []struct {
+		note      string
+		query     string
+		hit, miss uint64
+		exp       interface{} // if non-nil, check var `x`
+	}{
+		{
```

Comment:
Cache invalidation should be handled already but an extra test case would not hurt.

```
f(x) = y { x+input = y }
```

Query:

```
f(1, z) with input as 7; f(1, z2) with input as 8
```